### PR TITLE
HPCC-22627 Add skew info to logical files when published

### PR DESCRIPTION
**Pull request by jakesmith**
Expose on file details page.

Signed-off-by: Jake Smith <jake.smith@lexisnexisrisk.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by jakesmith** [comment]
@wangkx - please review.

Do dfu_file.xslt (there's 2 of them) need changing as well?
I did alter them, but they seemed to make no difference to eclwatch.

File: esp/services/ws_dfu/ws_dfuService.cpp [hunk 1]
```diff
@@ -2351,6 +2351,28 @@ void CWsDfuEx::doGetFileDetails(IEspContext &context, IUserDescriptor *udesc, co
...
+            Stat.setMinSkewInt64(minSkew);
+            Stat.setMaxSkewInt64(maxSkew);
+        }
+        if (version >= 1.52)
+        {
+            Stat.setMinSkewPart(minSkewPart);
+            Stat.setMaxSkewPart(maxSkewPart);
+        }
+
+        VStringBuffer minSkewString("-%.2f", ((double)minSkew)/100);
```
Comment (wangkx) [inline esp/services/ws_dfu/ws_dfuService.cpp]: Should the minus be added if minSkew = 0?

**Review by wangkx — COMMENTED** [review]
@jakesmith Please check my 2 questions.

In ws_dfuService.cpp, you may also define out the getFilePartsOnClusters() because it is not needed anymore.

The dfu_file.xslt is for legacy ECLwatch.

File: dali/base/dadfs.cpp [hunk 2]
```diff
@@ -3358,6 +3358,58 @@ class CDistributedFile: public CDistributedFileBase<IDistributedFile>
...
+            if (size < minPartSz)
+            {
+                minPartSz = size;
+                minSkewPart = p;
+            }
+            totalPartSz += size;
+        }
+        offset_t avgPartSz = totalPartSz / np;
+
+        maxSkew = (unsigned)(10000.0 * (((double)maxPartSz-avgPartSz)/avgPartSz));
```
[the remark this answers is not in the document]
Comment (jakesmith) [reply dali/base/dadfs.cpp]: To keep it as an unsigned with 2 point precision. When it is divided by 100 it may become e.g. 123.45%

**Comment by jakesmith** [comment]
>The dfu_file.xslt is for legacy ECLwatch.

@wangkx - should they be kept updated?

**Comment by jakesmith** [comment]
>In ws_dfuService.cpp, you may also define out the getFilePartsOnClusters() because it is not needed anymore.

I did not mean to leave defined out, I meant to delete, but thanks, yes I'll also delete getFilePartsOnClusters.
Can you double check nothing in the code being removed is needed?

**Comment by wangkx** [comment]
If you can update the dfu_file.xslt, it will be great. We do not support the legacy ECLWatch officially. But, someone still uses it.

I haven't found any code being removed is needed.

**Comment by jakesmith** [comment]
@wangkx - have deleted the defined out code and getFilePartsOnClusters()
After speaking to Gordon, I won't change legacy files. If people want the new stuff they shouldn't be using legacy.

Please review.

File: esp/services/ws_dfu/ws_dfuService.cpp (hunk 1 above)
Comment (wangkx) [reply esp/services/ws_dfu/ws_dfuService.cpp]: @jakesmith did you miss this comment?

Comment (jakesmith) [reply esp/services/ws_dfu/ws_dfuService.cpp]: I did miss it, thanks. I agree should be removed.
I'll change.

Comment (jakesmith) [reply esp/services/ws_dfu/ws_dfuService.cpp]: @wangkx - have now amended to avoid adding minus if no skew.

**Review by wangkx — APPROVED** [review]
@jakesmith looks fine.

**Comment by jakesmith** [comment]
@richardkchapman - please merge.

**Comment by richardkchapman** [comment]
@jakesmith Should this target 7.6?

**Comment by jakesmith** [comment]
@richardkchapman - yeah think it should, have rebased

**Pull request merged**